### PR TITLE
Rewrite JavaVersion

### PR DESCRIPTION
**Pull request by blazember**
Java9 introduced `java.lang.Runtime.Version` with a `major()` method that returns the major version as an int that the replaced code tried to figure out by parsing the `java.version` system property. Since with 4.0 Hazelcast requires at least Java8 runtime, it's easy to detect if the runtime is Java8: this class is not accessible via reflection. Otherwise, we can read out the major version via the mentioned method. Since this is part of the official API dealing with unknown Java versions is not an issue anymore, we can safely compare expected versions against the runtime version. For this purpose `JavaVersion.FutureJavaVersion` inner class and the `JavaMajorVersion` interface is introduced.

In the scope of this work, `ClassloadingMutexProvider` got removed. This provider checked if parallel class loading was in use (if Java7+ version) and returned a mutex safe to be used with the class loading mechanism. With Hazelcast 4.0 and Java8 we always take the Java7+ branch, therefore no need for this class.

Implements #15348

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/internal/util/JavaVersion.java [hunk 1]
```diff
@@ -53,44 +77,83 @@ public static boolean isAtMost(JavaVersion version) {
...
 
-    static boolean isAtLeast(JavaVersion currentVersion, JavaVersion minVersion) {
-        return currentVersion.ordinal() >= minVersion.ordinal();
+    private static JavaMajorVersion detectCurrentVersion() {
+        final ILogger logger = Logger.getLogger(JavaVersion.class);
+        final Class runtimeClass;
+        final Class versionClass;
+
+        try {
+            runtimeClass = Class.forName("java.lang.Runtime");
```
Comment (kwart) [inline hazelcast/src/main/java/com/hazelcast/internal/util/JavaVersion.java]: We don't need reflection here, I think. The `Runtime` is part of Java from its beginning. `Runtime.class` should do the job.

File: hazelcast/src/main/java/com/hazelcast/internal/util/JavaVersion.java [hunk 2]
```diff
@@ -53,44 +77,83 @@ public static boolean isAtMost(JavaVersion version) {
...
+        } catch (ClassNotFoundException e) {
+            logger.warning("Unable to detect Java version, falling back to Java 8", e);
+            return JAVA_8;
+        }
+
+        try {
+            versionClass = Class.forName("java.lang.Runtime$Version");
+        } catch (ClassNotFoundException e) {
+            // if Runtime is present but Runtime.Version doesn't, it's Java8 for sure
+            logger.info("Detected runtime version: Java 8");
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/internal/util/JavaVersion.java]: reduce logging level to `fine` here?

File: hazelcast/src/main/java/com/hazelcast/internal/util/JavaVersion.java (hunk 1 above)
Comment (blazember) [reply hazelcast/src/main/java/com/hazelcast/internal/util/JavaVersion.java]: Oops. Of course :) Fixed.

**Comment by blazember** [comment]
Thanks for the review @kwart @vbekiaris :+1:

**Pull request merged**